src/nvim/ex_cmds.c

```diff
@@ -5846,3 +5904,307 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+    char *str = xcalloc((size_t) curwin->w_frame->fr_width, sizeof(char));
+
+    // Append the lines to our buffer
+    kl_iter(matchedline_T, lmatch, current) {
+      matchedline_T mat = (*current)->data;
+      size_t line_size = STRLEN(mat.line) + col_width + 1;
+
+      // Reallocation if str not long enough
+      if (line_size > curwin->w_frame->fr_width * sizeof(char)) {
+        str = xrealloc(str, line_size * sizeof(char));
+      }
+
+      // Add the line number to the string
+      char *col = compute_line_number(col_width, mat.lnum);
+      snprintf(str, line_size, "%s%s", col, mat.line);
+      ml_append(line++, (char_u *) str, (colnr_T) 0, false);
+
+      int i = 0;
+      kl_iter(colnr_T, mat.start_col, col) {
+        // highlight the replaced part
+        if (sub_size > 0) {
+          src_id_highlight =
+            bufhl_add_hl(curbuf,
+                         src_id_highlight,
+                         2,  // id of our highlight
+                         line,
+                         (*col)->data + col_width + i * (sub_size - pat_size) + 1,
+                         (*col)->data + col_width + i * (sub_size - pat_size) +
+                         sub_size);
+        }
+        i++;
+      }
+
+      // free of the saved line and the allocated column
+      xfree(col);
+      xfree((*current)->data.line);
+    }
+    xfree(str);
+  }
+
+  redraw_later(SOME_VALID);
+
+  // Restore the old window
+  win_enter(oldwin, false);
+  finish_live_cmd(save_State, &winsizes, save_exmode,
+                  save_restart_edit, save_cmdmsg_rl, 1);
+
+  return;
+}
+
+/// Parse the substitution command line
+//
+/// @param eap arguments of the substitution
+/// @return cmdl_progress
+/// @see LiveSub_state definition
+LiveSub_state parse_sub_cmd(exarg_T *eap) {
+  int i = 0;
+  LiveSub_state cmdl_progress;
+
+  if (eap->arg[i++] != '/') {
+    return LS_NO_WD;
+  }
+
+  if (eap->arg[i++] == 0) {
+    return LS_NO_WD;
+  } else {
+    cmdl_progress = LS_ONE_WD;
+    while (eap->arg[i] != 0) {
+      if (eap->arg[i] == '/' && eap->arg[i-1] != '\\') {
+        cmdl_progress = (eap->arg[i+1] == 0) ? LS_TWO_SLASH_ONE_WD : LS_TWO_WD;
+        break;
+      }
+      i++;
+    }
+  }
+  return cmdl_progress;
+}
+
+/// This function is called when CMD_SUBSTITUTE is detected
+/// It will then proceed to launch do_sub() and ':u'
+/// at every new character typed in the cmdbuff according to the
+/// actual state of the live_substitution
+void do_live_sub(exarg_T *eap) {
+  // if livesub disabled, do it the classical way
+  if (!p_sub) {
+    do_sub(eap);
+    return;
+  }
+
+  // count the number of '/' to know how many words can be parsed
+  LiveSub_state cmdl_progress = parse_sub_cmd(eap);
+
+  char_u *arg;
+  char_u *tmp;
+
+  switch (cmdl_progress) {
+    case LS_NO_WD:
+      if (!EVENT_COLON) {
+        do_sub(eap);
+      }
+      break;
+    case LS_ONE_WD:
+      if (EVENT_SUB == 1 && sub_done == 1) {
+        do_cmdline_cmd(":u");
```

Comment:
I wonder how hard it would be to implement an internal function to "undo and pretend the last edit never happened" to not mess up `g+`. But then again I think it would be better to change the text that `screen.c` draws without actually changing the buffer. I did basic test for this when this feature work started, changing part of a line (multi-char conceal if you want) was not very hard except for getting the cursor position right, but the cursor is not inside the buffer for this usecase anyway...